src/serialization.jl

```diff
@@ -57,8 +63,59 @@ function serialize(vals::Vector{T}) where {T <: InfrastructureSystemsType}
     return serialize_struct.(vals)
 end
 
+function serialize(func::Function)
+    return Dict{String, Any}(
+        METADATA_KEY => Dict{String, Any}(
+            FUNCTION_KEY => string(nameof(func)),
+            MODULE_KEY => string(parentmodule(func)),
+        ),
+    )
+end
+
 function serialize_struct(val::T) where {T}
-    return Dict(string(name) => serialize(getfield(val, name)) for name in fieldnames(T))
+    @debug "serialize_struct" val T
+    data = Dict(string(name) => serialize(getfield(val, name)) for name in fieldnames(T))
+    add_serialization_metadata!(data, T)
+    return data
+end
+
+"""
+Add type information to the dictionary that can be used to deserialize the value.
+"""
+function add_serialization_metadata!(data::Dict, ::Type{T}) where {T}
+    data[METADATA_KEY] = Dict{String, Any}(
+        TYPE_KEY => string(nameof(T)),
+        MODULE_KEY => string(parentmodule(T)),
+    )
+    if !isempty(T.parameters)
+        data[METADATA_KEY][PARAMETERS_KEY] = string.(T.parameters)
+    end
+
+    return
+end
+
+"""
+Return the type information for the serialized struct.
+"""
+get_serialization_metadata(data::Dict) = data[METADATA_KEY]
+
+function get_type_from_serialization_data(data::Dict)
+    return get_type_from_serialization_metadata(get_serialization_metadata(data))
+end
+
+function get_type_from_serialization_metadata(metadata::Dict)
+    _module = Base.root_module(Base.__toplevel__, Symbol(metadata[MODULE_KEY]))
+    base_type = getfield(_module, Symbol(metadata[TYPE_KEY]))
+    if !get(metadata, CONSTRUCT_WITH_PARAMETERS_KEY, false)
+        return base_type
+    end
+
+    # This has several limitations and is only a workaround for PSY.Reserve subtypes.
+    # - each parameter must be in _module
+    # - does not support nested parametrics.
+    # Reserves should be fixed and then we can remove this hack.
+    parameters = [getfield(_module, Symbol(x)) for x in metadata[PARAMETERS_KEY]]
+    return base_type{parameters...}
```

Comment:
@jd-lara You have said that you want to remove parameterization from VariableReserve.  Doing so would require changes in PSI, PSY, and here.  Do you want to do that for v1.0?

Reply:
we should probably leave it as it is. It is more complex than we need but there isn't anything really broken. We can stage this for the following breaking change.